# FragFlow
*FragFlow: Automated Workflow for Large-Scale Quantitative Proteomics in High Performance Computing Environments*

### Workflow
![image](./workflow.png)
**Figure - Overview of the nf-fphpc workflow.** The workflow is composed of five sub-workflows: **manifest** file generation part creates the first input file based on a specified folder, with an optional MS raw file conversion step; **database** module downloads the reference proteome and augments them with decoy sequences; a **workflow** module which defines the desired analysis mode provided by FP; **FragPipe** module performs the actual MS analysis; **FragPipe-Analyst** module conducts the downstream post-processing analysis. Our workflow supports all predefined FragPipe analysis modes (FragPipe workflows) as well as custom user-defined configurations.
### Introduction
lorem ipsum

-----

## Installation
### Nextflow
The only requirement is the installation of Nextflow, which can be easily performed using their self-installing package. Follow the official website for more information: [Installation - Nextflow](https://www.nextflow.io/docs/latest/install.html)
``` 
curl -s https://get.nextflow.io | bash
```
## FragFlow Analysis
### 1. Preparing input files
Currently, by default ```.raw``` and ```.mzML``` mass spectrometry files are supported, although other file types can also be used, see MSConverter sub-workflow.\
Please prepare the input files using FragPipe manifest generation guidelines e.g. ```<filename>_<experiment>_<bioreplicate>_<mode>.raw```, where\
- filename - is the original filename of the raw MS file
- experiment - which is the group, experiment, treatment identifier of the sample (can be text, number, or empty)
- bioreplicate - which is the numerical identifier of a sample (can be number, or empty)
- mode - which is the pre-defined data types for aqcuisition modes used in FragPipe (DDA, DDA+, DIA, DIA-Quant, or DIA-lib)\

**Examples for file names:** \
- *filename_control_1_DDA.mzML*
- *filename_KO_3.raw*
- *filename_treatment_DIA.raw*
- *filename_control.mzML*

### 2. Configuration
Input parameters can be defined in multiple ways:
- A. modifying the ```nextflow.config``` file (recommended only when pulling the repository)
- B. custom config file ```<your_file>.config``` in a ```YAML``` or ```JSON``` format defined during run
```
nextflow run sznistvan/nf-fphpc -c <your_file>.config
```
- C. Defining parameters during run, see Parameters
```
nextflow run sznistvan/nf-fphpc \
    --input_folder "/path/to/input/folder" \
    --ram 32 \
    --threads 16 \
    --workflow "LFQ-MBR" \
    --mode "DDA" \
    --fasta_file "UP000005640"
```

The mandatory parameters for execution are the ```--input_folder```, ```--workflow```, ```--mode```, and ```--fasta_file```. See Parameters for more details.

### 3. Running FragFlow
When runnning FragFlow, all output files will be generated into the current working directory. If multiple analysis are performed, **please be cautious about current working directroy before running the analysis!**
``` 
nextflow run sznistvan/nf-fphpc
```
For long analysis times, it is benefitial to start FragFlow as a background process, which allows user interactions with the terminal and session logouts. For this, the ```-bg``` option should be added to the run call, and redirect the *stdout* to a log file.
```
nextflow run sznistvan/nf-fphpc -bg > fragflow.log
```
In order to cancel the currently running background process please use ```ps``` and/or the ID in ```.nextflow.pid``` to ```kill``` the process. More information.

### 4. Config Tools Download
To enable user-friendly execution, the download process of MSFragger, diaTracer, and IonQuant tools are also automated. For the initial analysis, all config tools will be downloaded keeping all necessary licensing directives nice. 

### 5. Notes on DIA-NN version
Currently FragPipe supports DIA-NN version 1.8.2beta. With FragFlow, similarly to the GUI version of FP, users can add custom versions of DIA-NN to use. https://github.com/vdemichev/DiaNN/releases/latest
Please provide the download link to the *Linux* version of any DIA-NN release using the ```--diann_download``` parameter.\
*Note: currently when using MSBooster, the newer versions of DIA-NN does not run successfully. For this reason, when specifying newer version of DIA-NN (1.8.2+), the newer version will only be used during the actual DIA-NN analysis. Additionally, current changes in newer DIA-NN modules, can raise errors in FP, which will be eventually handled by FP developers in future releases. This does not effect the results generated by FragFlow.*

### 6. Parameters
Parameters are accesible through the *nextflow.config* file. Additionally, every parameter can be added during execution e.g. *--input_folder*.

The following table lists all available parameters.
**TABLE**

## Individual Sub-Workflows
### MSConverter
adas
### FragPipe
run only fragpipe

### FragPipe-Analyst
run only fragpipe anal

## References
- FragPipe
- FragPipe-Analyst
- Ionquant, MSFragger, diaTracer
- DIA-NN

## Authors
<ins>Istvan Szepesi-Nagy</ins>(1,6), Roberta Borosta(1,7), Zoltan Szabo(2), Gabor Tusnady(3), Lorinc Pongor(4), Gergely Rona(1,5)

1. DNA Repair Research Group, Institute of Molecular Life Sciences, HUN-REN Research Centre for Natural Sciences, Budapest, HU
2. Department of Medical Chemistry, Albert Szent-Györgyi Medical School, University of Szeged, Szeged, HU
3. Proteomics Bioinformatics Research Group, Institute of Molecular Life Sciences, HUN-REN Research Centre for Natural Sciences, Budapest, HU
4. Cancer Genomics and Epigenetics Core Group, Hungarian Centre of Excellence for Molecular Medicine (HCEMM), Szeged, HU
5. Department of Biochemistry and Molecular Pharmacology, NYU Grossman School of Medicine, NY, USA
6. Semmelweis University Doctoral School, Budapest, HU
7. Doctoral School of Biology, ELTE Eötvös Loránd University, Budapest, HU
